rs/bitcoin/kyt/src/state.rs

```diff
@@ -52,37 +52,107 @@ pub struct FetchedTx {
 // Max number of concurrent http outcalls.
 const MAX_CONCURRENT: u32 = 50;
 
+// Default expiry of cached fetch transaction entries is 7 days (in nano seconds).
+const DEFAULT_FETCH_TX_EXPIRY: u64 = 7 * 24 * 3600 * 1000000;
+
+// Max number of entries in the cache is set to 10_000. Since the average transaction size
+// is about 400 bytes, the estimated memory usage of the cache is in the order of 10s of MBs.
+const MAX_FETCH_TX_ENTRIES: usize = 10_000;
+
 // The internal KYT state includes:
 // 1. Outcall capacity, a semaphore limiting max concurrent outcalls.
 // 2. fetch transaction status, indexed by transaction id.
 //
 // TODO(XC-191): persist canister state
 thread_local! {
     static OUTCALL_CAPACITY: RefCell<u32> = const { RefCell::new(MAX_CONCURRENT) };
-    static FETCH_TX_STATUS: RefCell<BTreeMap<Txid, FetchTxStatus>> = RefCell::new(BTreeMap::default());
+    static FETCH_TX_CACHE: RefCell<FetchTxCache<FetchTxStatus>> = RefCell::new(
+        FetchTxCache::new(DEFAULT_FETCH_TX_EXPIRY, MAX_FETCH_TX_ENTRIES)
+    );
+}
+
+// Time in nanoseconds since the epoch (1970-01-01).
+type Timestamp = u64;
+
+// Time duration in nanoseconds.
+type Duration = u64;
+
+struct FetchTxCache<T> {
+    expiry: Duration,
+    max_entries: usize,
+    status: BTreeMap<Txid, T>,
+    created: VecDeque<(Txid, Timestamp)>,
```

Comment:
understanding question: have you maybe thought of keeping a single data structure for `status` and `created`, e.g.,
```rust
status: BTreeMap<Txid, (Timestamp, T)>
```
Right now it seems that `max_entries` refers to the size of `created` but `created` could contain several entries for a given `Txid`, so that it's actually not immediately clear how many `Txid` the cache could store.

Reply:
The reason to have a separate `DeQueue` (which is effectively just a LRU cache) is to have efficient purging. Otherwise purge by timestamp would be a traversal.

The implementation also ensures that the number of entries in both the map and dequeue are always kept the same, but like you said, it is not obvious.

Reply:
> The implementation also ensures that the number of entries in both the map and dequeue are always kept the same

ha true I forgot that `created` is not modified when we set the status of an already present `Txid`.

Maybe it would make sense to have a unit test for that, e.g.
```rust
#[test]
fn should_have_same_size() {
    enum Status {
        Status0,
        Status1,
        Status2,
    }

    let txid_0 = Txid::from([0u8; 32]);
    let max_entries = 3;
    let mut cache = FetchTxCache::new(max_entries);
    assert_eq!(cache.status.len(), 0);
    assert_eq!(cache.created.len(), 0);

    cache.set_status_with(txid_0, Status::Status0, 0);
    assert_eq!(cache.status.len(), 1);
    assert_eq!(cache.created.len(), 1);

    cache.set_status_with(txid_0, Status::Status1, 1);
    assert_eq!(cache.status.len(), 1);
    assert_eq!(cache.created.len(), 1);

    cache.set_status_with(txid_0, Status::Status2, 2);
    assert_eq!(cache.status.len(), 1);
    assert_eq!(cache.created.len(), 1);
}
```

Reply:
Thanks! Added to tests